builder/dockerfile/internals.go

```diff
@@ -0,0 +1,688 @@
+	"github.com/docker/docker/pkg/archive"
+	"github.com/docker/docker/pkg/httputils"
+	"github.com/docker/docker/pkg/stringid"
+	"github.com/docker/docker/pkg/stringutils"
+	"github.com/docker/docker/pkg/system"
+	"github.com/docker/docker/pkg/tarsum"
+	"github.com/docker/docker/pkg/urlutil"
+	"github.com/docker/docker/runconfig"
+)
+
+func (b *Builder) commit(id string, autoCmd *stringutils.StrSlice, comment string) error {
+	if b.disableCommit {
+		return nil
+	}
+	if b.image == "" && !b.noBaseImage {
+		return fmt.Errorf("Please provide a source image with `from` prior to commit")
+	}
+	b.runConfig.Image = b.image
+	if id == "" {
+		cmd := b.runConfig.Cmd
+		if runtime.GOOS != "windows" {
+			b.runConfig.Cmd = stringutils.NewStrSlice("/bin/sh", "-c", "#(nop) "+comment)
+		} else {
+			b.runConfig.Cmd = stringutils.NewStrSlice("cmd", "/S /C", "REM (nop) "+comment)
+		}
+		// TODO: weird
+		defer func(cmd *stringutils.StrSlice) { b.runConfig.Cmd = cmd }(cmd)
+
+		if hit, err := b.probeCache(); err != nil {
+			return err
+		} else if hit {
+			return nil
+		}
+
+		container, err := b.create()
+		if err != nil {
+			return err
+		}
+		id = container.ID
+
+		if err := container.Mount(); err != nil {
+			return err
+		}
+		defer container.Unmount()
+	}
+
+	//container, err := b.Daemon.Get(id)
+	container, err := b.docker.Container(id)
+	if err != nil {
+		return err
+	}
+
+	// Note: Actually copy the struct
+	autoConfig := *b.runConfig
+	autoConfig.Cmd = autoCmd
+
+	commitCfg := &daemon.ContainerCommitConfig{
+		Author: b.maintainer,
+		Pause:  true,
+		Config: &autoConfig,
+	}
+
+	// Commit the container
+	//image, err := b.Daemon.Commit(container, commitCfg)
+	image, err := b.docker.Commit(container, commitCfg)
+	if err != nil {
+		return err
+	}
+	// TODO: transaction
+	//b.Daemon.Graph().Retain(b.id, image.ID)
+	//b.activeImages = append(b.activeImages, image.ID)
+	b.image = image.ID
+	return nil
+}
+
+type copyInfo struct {
+	os.FileInfo
+	decompress bool
+}
+
+func (b *Builder) runContextCommand(args []string, allowRemote bool, allowLocalDecompression bool, cmdName string) error {
+	if b.context == nil {
+		return fmt.Errorf("No context given. Impossible to use %s", cmdName)
+	}
+
+	if len(args) < 2 {
+		return fmt.Errorf("Invalid %s format - at least two arguments required", cmdName)
+	}
+
+	// Work in daemon-specific filepath semantics
+	dest := filepath.FromSlash(args[len(args)-1]) // last one is always the dest
+
+	b.runConfig.Image = b.image
+
+	var infos []copyInfo
+
+	// Loop through each src file and calculate the info we need to
+	// do the copy (e.g. hash value if cached).  Don't actually do
+	// the copy until we've looked at all src files
+	var err error
+	for _, orig := range args[0 : len(args)-1] {
+		var fi os.FileInfo
+		decompress := allowLocalDecompression
+		if urlutil.IsURL(orig) {
+			if !allowRemote {
+				return fmt.Errorf("Source can't be a URL for %s", cmdName)
+			}
+			fi, err = b.download(orig)
```

Comment:
@LK4D4 no. I addressed it in a previous version and had to revert because it complicates things a lot. I can add a `TODO` for later fix. It should get simplified once `builder.PathFileInfo` etc. and all those are removed.